File: HDF5Examples/JAVA/H5G/H5Ex_G_Traverse.java
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Copyright by The HDF Group.                                               *
 * All rights reserved.                                                      *
 *                                                                           *
 * This file is part of HDF5.  The full HDF5 copyright notice, including     *
 * terms governing use, modification, and redistribution, is contained in    *
 * the LICENSE file, which can be found at the root of the source code       *
 * distribution tree, or in https://www.hdfgroup.org/licenses.               *
 * If you do not have access to either file, you may request a copy from     *
 * help@hdfgroup.org.                                                        *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

/************************************************************
This example shows a way to recursively traverse the file
using H5Literate.  The method shown here guarantees that
the recursion will not enter an infinite loop, but does
not prevent objects from being visited more than once.
The program prints the directory structure of the file
specified in FILE.  The default file used by this example
implements the structure described in the User's Guide,
chapter 4, figure 26.
 ************************************************************/

import static org.hdfgroup.javahdf5.hdf5_h.*;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.ArrayList;
import java.util.List;

import hdf.hdf5lib.structs.H5O_token_t;

import org.hdfgroup.javahdf5.*;

public class H5Ex_G_Traverse {

    private static String FILENAME = "groups/h5ex_g_traverse.h5";

    // Operator data structure for iteration
    static class OpData {
        int recurs;            // Recursion level
        OpData prev;           // Previous operator data
        H5O_token_t obj_token; // Object token
    }

    private static void OpenGroup(Arena arena)
    {
        long file_id = H5I_INVALID_HID();

        try {
            // Open file
            file_id = H5Fopen(arena.allocateFrom(FILENAME), H5F_ACC_RDONLY(), H5P_DEFAULT());

            if (file_id >= 0) {
                // Get info for root group
                MemorySegment root_info = arena.allocate(H5O_info2_t.sizeof());
                H5Oget_info3(file_id, root_info, H5O_INFO_ALL());

                // Initialize operator data
                OpData od    = new OpData();
                od.recurs    = 0;
                od.prev      = null;
                od.obj_token = new H5O_token_t(H5O_info2_t.token(root_info));

                // Print root group and begin iteration
                System.out.println("/ {");

                // Create and iterate
                iterateGroup(arena, file_id, od);

                System.out.println("}");
            }
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        finally {
            // Close and release resources
            try {
                if (file_id >= 0)
                    H5Fclose(file_id);
            }
            catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static void iterateGroup(Arena arena, long group_id, OpData od)
    {
        // Create callback for H5Literate
        H5L_iterate2_t.Function link_callback =
            (long group, MemorySegment name, MemorySegment info, MemorySegment op_data) ->
        {
            String link_name = name.getString(0);
            int spaces       = 2 * (od.recurs + 1); // Indentation

            try {
                // Get object info
                MemorySegment obj_info = arena.allocate(H5O_info2_t.sizeof());
                int ret = H5Oget_info_by_name3(group, name, obj_info, H5O_INFO_ALL(), H5P_DEFAULT());

                if (ret >= 0) {
                    int obj_type          = H5O_info2_t.type(obj_info);
                    H5O_token_t obj_token = new H5O_token_t(H5O_info2_t.token(obj_info));

                    // Print indentation
                    for (int i = 0; i < spaces; i++)
                        System.out.print(" ");

                    if (obj_type == H5O_TYPE_GROUP()) {
                        System.out.println("Group: " + link_name + " {");

                        // Check for loops
                        if (groupCheck(od, obj_token)) {
                            for (int i = 0; i < spaces; i++)
                                System.out.print(" ");
                            System.out.println("  Warning: Loop detected!");
                        }
                        else {
                            // Create new operator data for recursion
                            OpData nextod    = new OpData();
                            nextod.recurs    = od.recurs + 1;
                            nextod.prev      = od;
                            nextod.obj_token = obj_token;

                            // Recurse into group
                            long subgroup_id = H5Gopen2(group, name, H5P_DEFAULT());
                            if (subgroup_id >= 0) {
                                try {
                                    iterateGroup(arena, subgroup_id, nextod);
                                }
                                finally {
                                    H5Gclose(subgroup_id);
                                }
                            }
                        }

                        for (int i = 0; i < spaces; i++)
                            System.out.print(" ");
                        System.out.println("}");
                    }
                    else if (obj_type == H5O_TYPE_DATASET()) {
                        System.out.println("Dataset: " + link_name);
                    }
                    else if (obj_type == H5O_TYPE_NAMED_DATATYPE()) {
                        System.out.println("Datatype: " + link_name);
                    }
                    else {
                        System.out.println("Unknown: " + link_name);
                    }
                }
            }
            catch (Exception e) {
                e.printStackTrace();
            }

            return 0; // Continue iteration
        };

        // Allocate upcall stub
        MemorySegment link_callback_stub = H5L_iterate2_t.allocate(link_callback, arena);

        // Iterate over links in group
        H5Literate2(group_id, H5_INDEX_NAME(), H5_ITER_NATIVE(), MemorySegment.NULL, link_callback_stub,
                    MemorySegment.NULL);
    }

    // Check if we've already visited this object token (loop detection)
    private static boolean groupCheck(OpData od, H5O_token_t target_token)
    {
        if (od.obj_token.equals(target_token))
            return true; // Object tokens match
        else if (od.recurs == 0)
            return false; // Root group reached with no matches
        else
            return groupCheck(od.prev, target_token); // Recursively examine the next node
    }

    public static void main(String[] args)
    {
        try (Arena arena = Arena.ofConfined()) {
            H5Ex_G_Traverse.OpenGroup(arena);
        }
    }
}

File: HDF5Examples/C/H5G/h5ex_g_traverse.c
/************************************************************

  This example shows a way to recursively traverse the file
  using H5Literate.  The method shown here guarantees that
  the recursion will not enter an infinite loop, but does
  not prevent objects from being visited more than once.
  The program prints the directory structure of the file
  specified in FILE.  The default file used by this example
  implements the structure described in the User's Guide,
  chapter 4, figure 26.

Note: This example includes older cases from previous versions
  of HDF5 for historical reference and to illustrate how to
  migrate older code to newer functions. However, readers are
  encouraged to avoid using deprecated functions and earlier
  schemas from those versions.
 ************************************************************/

#include "hdf5.h"
#include <stdio.h>
#include <string.h>

#define FILENAME "h5ex_g_traverse.h5"

/*
 * Define operator data structure type for H5Literate callback.
 * During recursive iteration, these structures will form a
 * linked list that can be searched for duplicate groups,
 * preventing infinite recursion.
 */
struct opdata {
    unsigned       recurs; /* Recursion level.  0=root */
    struct opdata *prev;   /* Pointer to previous opdata */
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
    H5O_token_t token; /* Group token */
#else
    haddr_t addr; /* Group address */
#endif
};

/*
 * Operator function to be called by H5Literate.
 */
herr_t op_func(hid_t loc_id, const char *name, const H5L_info_t *info, void *operator_data);

#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
/* Function to check for duplicate groups in a path. */
int group_check(hid_t loc_id, struct opdata *od, H5O_token_t target_token);

#else
/* Function to check for duplicate groups in a path. */
int group_check(struct opdata *od, haddr_t target_addr);

#endif

int
main(void)
{
    hid_t         file; /* Handle */
    herr_t        status;
    H5O_info_t    infobuf;
    struct opdata od;

    /*
     * Open file and initialize the operator data structure.
     */
    file = H5Fopen(FILENAME, H5F_ACC_RDONLY, H5P_DEFAULT);
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
    status = H5Oget_info(file, &infobuf, H5O_INFO_ALL);
    memcpy(&od.token, &infobuf.token, sizeof(H5O_token_t));
#else
    status  = H5Oget_info(file, &infobuf);
    od.addr = infobuf.addr;
#endif
    od.recurs = 0;
    od.prev   = NULL;

    /*
     * Print the root group and formatting, begin iteration.
     */
    printf("/ {\n");
    status = H5Literate(file, H5_INDEX_NAME, H5_ITER_NATIVE, NULL, op_func, (void *)&od);
    printf("}\n");

    /*
     * Close and release resources.
     */
    status = H5Fclose(file);

    return 0;
}

/************************************************************

  Operator function.  This function prints the name and type
  of the object passed to it.  If the object is a group, it
  is first checked against other groups in its path using
  the group_check function, then if it is not a duplicate,
  H5Literate is called for that group.  This guarantees that
  the program will not enter infinite recursion due to a
  circular path in the file.

 ************************************************************/
herr_t
op_func(hid_t loc_id, const char *name, const H5L_info_t *info, void *operator_data)
{
    herr_t         status;
    herr_t         return_val = 0;
    struct opdata *od         = (struct opdata *)operator_data; /* Type conversion */
    unsigned       spaces     = 2 * (od->recurs + 1); /* Number of whitespaces to prepend to output */

    /*
     * Get type of the object and display its name and type.
     * The name of the object is passed to this function by
     * the Library.
     */
    H5O_info_t infobuf;
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
    status = H5Oget_info_by_name(loc_id, name, &infobuf, H5O_INFO_ALL, H5P_DEFAULT);
#else
    status = H5Oget_info_by_name(loc_id, name, &infobuf, H5P_DEFAULT);
#endif
    printf("%*s", spaces, ""); /* Format output */
    switch (infobuf.type) {
        case H5O_TYPE_GROUP:
            printf("Group: %s {\n", name);

            /*
             * Check group address/token against linked list of operator
             * data structures.  We will always run the check, as the
             * reference count cannot be relied upon if there are
             * symbolic links, and H5Oget_info_by_name always follows
             * symbolic links.  Alternatively we could use H5Lget_info
             * and never recurse on groups discovered by symbolic
             * links, however it could still fail if an object's
             * reference count was manually manipulated with
             * H5Odecr_refcount.
             */
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
            if (group_check(loc_id, od, infobuf.token)) {
#else
            if (group_check(od, infobuf.addr)) {
#endif
                printf("%*s  Warning: Loop detected!\n", spaces, "");
            }
            else {
                if (od->recurs + 1 > 7)
                    return -1;

                /*
                 * Initialize new operator data structure and
                 * begin recursive iteration on the discovered
                 * group.  The new opdata structure is given a
                 * pointer to the current one.
                 */
                struct opdata nextod;
                nextod.recurs = od->recurs + 1;
                nextod.prev   = od;
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
                memcpy(&nextod.token, &infobuf.token, sizeof(H5O_token_t));
#else
                nextod.addr = infobuf.addr;
#endif
                return_val = H5Literate_by_name(loc_id, name, H5_INDEX_NAME, H5_ITER_NATIVE, NULL, op_func,
                                                (void *)&nextod, H5P_DEFAULT);
            }
            printf("%*s}\n", spaces, "");
            break;
        case H5O_TYPE_DATASET:
            printf("Dataset: %s\n", name);
            break;
        case H5O_TYPE_NAMED_DATATYPE:
            printf("Datatype: %s\n", name);
            break;
        default:
            printf("Unknown: %s\n", name);
    }

    return return_val;
}

/************************************************************

  This function recursively searches the linked list of
  opdata structures for one whose address/token matches
  target_addr/token.  Returns 1 if a match is found, and 0
  otherwise.

 ************************************************************/
#if H5_VERSION_GE(1, 12, 0) && !defined(H5_USE_110_API) && !defined(H5_USE_18_API) && !defined(H5_USE_16_API)
int
group_check(hid_t loc_id, struct opdata *od, H5O_token_t target_token)
{
    int token_cmp;

    if (H5Otoken_cmp(loc_id, &od->token, &target_token, &token_cmp) < 0) {
        return 0;
    }
    if (!token_cmp) {
        return 1; /* Tokens match */
    }
    else if (!od->recurs)
        return 0; /* Root group reached with no matches */
    else {        /* Recursively examine the next node */
        return group_check(loc_id, od->prev, target_token);
    }
}
#else
int
group_check(struct opdata *od, haddr_t target_addr)
{
    if (od->addr == target_addr)
        return 1; /* Addresses match */
    else if (!od->recurs)
        return 0; /* Root group reached with no matches */
    else {        /* Recursively examine the next node */
        return group_check(od->prev, target_addr);
    }
}
#endif
